File: backend/app/Models/Transaction.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use App\Enums\TransactionCategory; // Add this line
use Illuminate\Database\Eloquent\Builder; // Add this line

class Transaction extends Model
{
    protected $fillable = [
        'user_id',
        'amount',
        'transaction_type',
        'category',
        'description',
        'attachment',
    ];

    protected $casts = [
        'category' => TransactionCategory::class, // Add this line
    ];

    protected static function boot()
    {
        parent::boot();

        static::addGlobalScope('orderByCreatedAt', function (Builder $builder) {
            $builder->orderBy('created_at', 'desc');
        });
    }


    public function user()
    {
        return $this->belongsTo(User::class);
    }

    public static function getAllTransactions()
    {
        return self::all();
    }


    // علاقة بالدخل
    public function income()
    {
        return $this->hasOne(Income::class);
    }

    // علاقة بالمصروف
    public function expense()
    {
        return $this->hasOne(Expense::class);
    }

    // جلب حسب نوع
    public static function getTransactionsByType($type)
    {
        return self::where('transaction_type', $type)->get();
    }

    // Add this method
    public static function getTransactionsByCategory(TransactionCategory $category)
    {
        return self::where('category', $category->value)->get();
    }

    // Add this method
    public function getCategoryLabel()
    {
        return TransactionCategory::getLabel($this->category);
    }

    public function scopeIncomes($query)
    {
        return $query->where('transaction_type', 'income');
    }

    public function scopeExpenses($query)
    {
        return $query->where('transaction_type', 'expense');
    }

    public function scopeIncomeCategory($query, string $category)
    {
        return $query->where('transaction_type', 'income')->where('category', $category);
    }

    public function scopeIncomeSum($query)
    {
        return $query->where('transaction_type', 'income')->sum('amount');
    }

    public function scopeExpenseSum($query)
    {
        return $query->where('transaction_type', 'expense')->sum('amount');
    }


    public function scopeIncomeCategorySum($query, TransactionCategory $category)
    {
        return $query->where('transaction_type', 'income')->where('category', $category)->sum('amount');
    }

    public function scopeExpenseCategorySum($query, TransactionCategory $category)
    {
        return $query->where('transaction_type', 'expense')->where('category', $category)->sum('amount');
    }

    public function scopeByCommittee($query, $committee)
    {
        return $query->whereHas('expense', function ($query) use ($committee) {
            $query->where('committee', $committee);
        });
    }
}

File: expence_app/lib/models/transaction.dart
import 'package:party_planner/enums/transaction_category.dart';
import 'package:party_planner/models/expense.dart';
import 'package:party_planner/models/income.dart';
import 'package:party_planner/models/user.dart';

class Transaction {
  String? id;
  String? amount;
  String? description;
  String? timeAgo;
  String? createdAt;
  String? updatedAt;
  TransactionCategory? category;
  String? attachment;
  UserModel? user;
  String? transactionType;
  dynamic _transaction;

  bool get isIncome => transactionType == 'income';

  Income get income => (_transaction as Income);
  Expense get expense => (_transaction);
  set transaction(dynamic value) => _transaction = value;

  Transaction({
    this.id,
    this.amount,
    this.description,
    this.createdAt,
    this.updatedAt,
    this.category,
    this.attachment,
    this.user,
    this.transactionType,
  });

  Transaction.fromJson(Map<String, dynamic> json) {
    id = json['id'].toString();
    amount = json['amount'].toString();
    description = json['description'];
    
    timeAgo = json['time_ago'];
    createdAt = json['formatDate'];

    updatedAt = json['updated_at'];
    category = TransactionCategory.values.firstWhere(
            (e) => e.toString() == "TransactionCategory.${json['category']}",
            orElse: () => TransactionCategory.transportation)
        as TransactionCategory?;
    attachment = json['attachment'];
    user = json['user'] != null ? UserModel.fromJson(json['user']) : null;
    transactionType = json['transaction_type'];
    transaction = transactionType == 'income'
        ? json['transaction'] != null
            ? Income.fromJson(json['transaction'])
            : null
        : json['transaction'] != null
            ? Expense.fromJson(json['transaction'])
            : null;
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = <String, dynamic>{};
    data['amount'] = amount;
    data['description'] = description;
    data['category'] = category?.toString().split('.').last;
    data['attachment'] = attachment;
    data['transaction_type'] = transactionType;
    
    
    if (user != null) {
      data['user'] = user!.toJson();
    }
    if (_transaction != null) {
      data['transaction'] =
          transactionType == 'income' ? income.toJson() : expense.toJson();
    }
    return data;
  }
}
